infra/gravity/cluster_failover.go

```diff
@@ -0,0 +1,118 @@
+package gravity
+
+import (
+	"context"
+
+	"github.com/gravitational/robotest/lib/wait"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+// Failover isolates the current leader node and elects a new leader node.
+// Conforms to ConfigFn interface.
+func (c *TestContext) Failover(nodes []Gravity) error {
+	// TODO: Configure timeouts
+	ctx, cancel := context.WithTimeout(c.ctx, c.timeouts.Status)
+	defer cancel()
+
+	oldLeader, err := getLeaderNode(ctx, nodes)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	c.Logger().WithFields(logrus.Fields{
+		"oldLeader": oldLeader,
+	}).Info("Leader found")
+
+	if err := oldLeader.PartitionNetwork(ctx); err != nil {
+		return trace.Wrap(err, "failed to create network partition")
+	}
+	// NOTE: is subnets an appropriate name for the separate networks, or is the
+	// term only relevant when grouping a range of IP addresses?
+	subnets := make([][]Gravity, 0, 2)
+	subnets[0] = []Gravity{oldLeader}
+	for i, node := range nodes {
+		if node == oldLeader {
+			subnets[1] = append(nodes[:i], nodes[i+1:]...)
+			break
+		}
+	}
+	c.Logger().WithFields(logrus.Fields{
+		"subnets": subnets,
+	}).Info("Created network partition")
+
+	retry := wait.Retryer{
+		Attempts: leaderElectionRetries,
+		Delay:    leaderElectionWait,
+	}
+
+	var newLeader Gravity
+	err = retry.Do(ctx, func() error {
+		newLeader, err = getLeaderNode(ctx, subnets[1])
+		if err != nil || newLeader == oldLeader {
+			return wait.Continue("new leader not yet elected", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return trace.Wrap(err, "new leader was not elected")
+	}
+
+	c.Logger().WithFields(logrus.Fields{
+		"oldLeader": oldLeader,
+		"newLeader": newLeader,
+	}).Info("New leader elected")
+
+	if err := oldLeader.UnpartitionNetwork(ctx); err != nil {
```

Comment:
Before restoring the network, should this run some tests on the second "partition" to make sure cluster is operational (e.g. gravity status, kubect get pods/nodes, etc.)?